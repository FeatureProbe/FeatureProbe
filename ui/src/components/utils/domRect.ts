export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}