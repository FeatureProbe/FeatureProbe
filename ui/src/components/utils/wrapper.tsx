import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
