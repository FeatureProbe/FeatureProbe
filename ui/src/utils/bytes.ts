export function getBytes(str: string) {
  let bytes = str.length,
    i = 0;
  for (; i < bytes; i++) if (str.charCodeAt(i) > 255) bytes++;

  return bytes;
}

const MB = 1048576;
export const MAX_SIZE = MB;