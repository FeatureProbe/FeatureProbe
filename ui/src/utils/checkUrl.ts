function escapeStringRegexp(string: string): string {
  if (typeof string !== 'string') {
    throw new TypeError('Expected a string');
  }

  return string
    .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
    .replace(/-/g, '\\x2d');
}

export function matchUrl(matcher: string, originUrl: string, targetUrl: string): boolean {
  let href = '';
  let hash = '';
  let search = '';

  try {
    const fullUrl = new URL(targetUrl, '');
    href = fullUrl.href;
    hash = fullUrl.hash;
    search = fullUrl.search;
  } catch (error) {
    href = targetUrl;
  }

  let regex;
  let testUrl;

  switch (matcher) {
    case 'EXACT':
      testUrl = href;
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SIMPLE':
      testUrl = href.replace(hash, '').replace(search, '');
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SUBSTRING':
      testUrl = href.replace(search, '');
      regex = new RegExp('.*' + escapeStringRegexp(originUrl) + '.*$');
      break;
    case 'REGULAR':
      testUrl = href.replace(search, '');
      regex = new RegExp(originUrl);
      break;
    default:
      return false;
  }

  return regex.test(testUrl);
}