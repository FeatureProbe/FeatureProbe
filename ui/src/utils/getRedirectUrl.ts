import { getProjectList } from 'services/project';
import { IProject, IEnvironment } from 'interfaces/project';
import { IDictionary } from 'interfaces/targeting';
import { getFromDictionary } from 'services/dictionary';
import { LAST_SEEN } from 'constants/dictionary_keys';

const getInfo = async () => {
  const res =  await getFromDictionary<IDictionary>(LAST_SEEN);
  const { success, data } = res;
  if (success && data) {
    return JSON.parse(data.value);
  } else {
    return {};
  }
};

export const getRedirectUrl = async (defaultRedirectUrl: string) => {
  let redirectUrl = defaultRedirectUrl;
  const info = await getInfo();
  const { projectKey, environmentKey } = info;

  if (!!projectKey && !!environmentKey) {
    redirectUrl = `/${projectKey}/${environmentKey}/toggles`;
  } else {
    const res = await getProjectList<IProject[]>();
    const { success, data } = res;
    if (success && data) {
      if (data[0]) {
        const { key, environments } = data[0];
        let envKey = '';
        environments.some((environment: IEnvironment) => {
          if (environment.key) {
            envKey = environment.key;
            return true;
          }
          return environment;
        });
  
        if (key && envKey) {
          redirectUrl = `/${key}/${envKey}/toggles`;
        }
      } else {
        redirectUrl = '/projects';
      }
    }
  }

  return redirectUrl;
};