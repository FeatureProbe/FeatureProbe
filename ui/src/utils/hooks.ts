
import { useState } from 'react';

export const useLocalStorage = (key: string, initialValue: boolean | string) => {
  const [storedValue, setStoredValue] = useState(() => {
    if (typeof window === 'undefined') {
      return initialValue;
    }

    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.log(error);
      return initialValue;
    }
  });

  const setValue = (value: object) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);

      if (typeof window !== 'undefined') {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
      
    } catch (error) {
      console.log(error);
    }
  };
  return [storedValue, setValue];
};
