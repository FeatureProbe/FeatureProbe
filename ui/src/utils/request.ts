function parseJSON(response: Response) {
  if (response.headers.get('token')) {
    localStorage.setItem('token', response.headers.get('token') || '');
  }
  return response.json();
}

function checkStatus(response: Response) {
  if (
    (response.status >= 200 && response.status < 300) 
    || response.status === 400 
    || response.status === 403
    || response.status === 404
  ) {
    return response;
  } 
  else if (response.status === 401) {
    window.location.href = '/login';
    return response;
  } 
  else {
    const error: Error = new Error(response.statusText);
    throw error;
  }
}

type IResponse<T> = {
  success: boolean;
  code?: string;
  message?: string;
  data?: T;
}

function handleCode<T, R>(res: IResponse<T>, resolve: (value: IResponse<R> | PromiseLike<IResponse<R>>) => void) {
  if (res.code) {
    resolve({
      success: false,
      code: res.code,
      message: res.message,
    });
  }
}

export default function request<T>(url: string, options?: RequestInit): Promise<IResponse<T>> {
  options = options || {
    headers:{
      'Accept-Language': 'en-US'
    }
  };

  return new Promise(resolve => {
    return fetch(url, options)
      .then(response => checkStatus(response))
      .then(response => parseJSON(response))
      .then(response => {
        handleCode<IResponse<T>, T>(response, resolve);
        resolve({
          success: true,
          data: response,
        });
      })
      .catch(err => {
        resolve({
          success: false,
          message: err.message,
        });
      });
  });
}
