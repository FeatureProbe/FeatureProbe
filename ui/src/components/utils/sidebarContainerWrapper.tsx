import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
