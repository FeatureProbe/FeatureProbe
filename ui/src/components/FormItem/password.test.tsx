import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { FieldErrorsImpl, useForm } from 'react-hook-form';
import { IntlWrapper } from 'components/utils/wrapper';
import FormItemPass from './password';

const SnapshotTest = (props: { errors?: Partial<FieldErrorsImpl<{
  [x: string]: unknown;
}>> }) => {
  const { register } = useForm();

  return <FormItemPass
    errors={props.errors ?? {}}
    register={register}
    onChange={() => {
      //
    }}
  />;
};

it('Password snapshot', () => {
  (async () => {
    const { asFragment, rerender } = render(
      <SnapshotTest />,
      {
        wrapper: IntlWrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();

    rerender(<SnapshotTest errors={{
      password: {
        message: 'error message',
        type: 'deps'
      }
    }} />);

    expect(asFragment()).toMatchSnapshot();
  })();
});
