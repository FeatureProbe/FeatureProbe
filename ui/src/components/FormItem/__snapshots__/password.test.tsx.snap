// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password snapshot 1`] = `
<DocumentFragment>
  <div
    class="field"
  >
    <label>
      <span
        class="label-required"
      >
        *
      </span>
      Password
    </label>
    <div
      class="field input"
    >
      <div
        class="ui input"
      >
        <input
          name="password"
          placeholder="Please input password"
          type="password"
        />
      </div>
    </div>
  </div>
  <div
    class="tip-text"
  >
    Password can only contain letters, numbers, '.', '_' or '-'.
  </div>
</DocumentFragment>
`;

exports[`Password snapshot 2`] = `
<DocumentFragment>
  <div
    class="field"
  >
    <label>
      <span
        class="label-required"
      >
        *
      </span>
      Password
    </label>
    <div
      class="error field input"
    >
      <div
        class="ui input"
      >
        <input
          aria-invalid="true"
          name="password"
          placeholder="Please input password"
          type="password"
        />
      </div>
    </div>
  </div>
  <div
    class="error-text"
  >
    error message
  </div>
  <div
    class="tip-text"
  >
    Password can only contain letters, numbers, '.', '_' or '-'.
  </div>
</DocumentFragment>
`;
